src/rnp/rnpcfg.c

```diff
@@ -0,0 +1,502 @@
+    if (!rnp_cfg_get_defkey(cfg, params))
+        return RNP_FAIL;
+
+    return RNP_OK;
+}
+
+/* find the value name in the rnp_cfg */
+static int
+rnp_cfg_find(rnp_cfg_t *cfg, const char *key)
+{
+    unsigned i;
+
+    for (i = 0; i < cfg->count && strcmp(cfg->keys[i], key) != 0; i++)
+        ;
+    return (i == cfg->count) ? -1 : (int) i;
+}
+
+/* resize keys/vals arrays to the new size. Only expanding is supported */
+static int
+rnp_cfg_resize(rnp_cfg_t *cfg, unsigned newsize)
+{
+    char **temp;
+
+    if (cfg->size == 0) {
+        /* only get here first time around */
+        cfg->keys = calloc(sizeof(char *), newsize);
+        cfg->vals = calloc(sizeof(char *), newsize);
+
+        if ((cfg->keys == NULL) || (cfg->keys == NULL)) {
+            (void) fprintf(stderr, "rnp_cfg_resize: bad alloc\n");
+            return RNP_FAIL;
+        }
+        cfg->size = newsize;
+    } else if (cfg->count == cfg->size) {
+        /* only uses 'needed' when filled array */
+        temp = realloc(cfg->keys, sizeof(char *) * newsize);
+        if (temp == NULL) {
+            (void) fprintf(stderr, "rnp_cfg_resize: bad realloc\n");
+            return RNP_FAIL;
+        }
+        cfg->keys = temp;
+
+        temp = realloc(cfg->vals, sizeof(char *) * newsize);
+        if (temp == NULL) {
+            (void) fprintf(stderr, "rnp_cfg_resize: bad realloc\n");
+            return RNP_FAIL;
+        }
+        cfg->vals = temp;
+        cfg->size = newsize;
+    }
+
+    return RNP_OK;
+}
+
+/* set val for the key in config. key and val are duplicated */
+int
+rnp_cfg_set(rnp_cfg_t *cfg, const char *key, const char *val)
+{
+    char *newval = NULL;
+    char *newkey;
+    int   i;
+
+    /* protect against the case where 'value' is rnp->value[i] */
+    if (val != NULL) {
+        newval = rnp_strdup(val);
+        if (newval == NULL) {
+            (void) fprintf(stderr, "rnp_cfg_set: bad alloc\n");
+            return RNP_FAIL;
+        }
+    }
+
+    if ((i = rnp_cfg_find(cfg, key)) < 0) {
+        /* add the element to the array */
+        if (rnp_cfg_resize(cfg, cfg->size + 15)) {
+            newkey = rnp_strdup(key);
+            if (newkey == NULL) {
+                (void) fprintf(stderr, "rnp_cfg_set: bad alloc\n");
+                return RNP_FAIL;
+            }
+            cfg->keys[i = cfg->count++] = newkey;
+        } else {
+            free(newval);
+            return RNP_FAIL;
+        }
+    } else {
+        /* replace the element in the array */
+        if (cfg->vals[i]) {
+            free(cfg->vals[i]);
+            cfg->vals[i] = NULL;
+        }
+    }
+
+    cfg->vals[i] = newval;
+    return RNP_OK;
+}
+
+/* unset var for key, setting it to NULL if it exists in cfg */
+int
+rnp_cfg_unset(rnp_cfg_t *cfg, const char *key)
+{
+    int i;
+
+    if ((i = rnp_cfg_find(cfg, key)) >= 0) {
+        free(cfg->vals[i]);
+        cfg->vals[i] = NULL;
+        return RNP_OK;
+    }
+    return RNP_FAIL;
```

Comment:
Return result can be used as indication whether key existed, maybe somewhere we'll make use of it.